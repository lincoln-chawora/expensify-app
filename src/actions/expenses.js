import db from "../firebase/firebase";

// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        // Create defaults, this is similar to adding them in arguments.
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;

        // Create expense object to store data in var.
        const expense = { description, note, amount, createdAt };

        // Add expense into firebase (push will generate id in fb), then dispatch that
        // expense into the redux store, using the new id from firebase.
        return db.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }))
        }).catch((e) => {
            console.log('Expense creation failed', e);
        })
    }
}
// Actions generators

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})

export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return db.ref(`users/${uid}/expenses/${id}`).remove().then( () => {
            dispatch(removeExpense({id}))
        }).catch((e) => {
            console.log('Unable to remove expenses', e);
        });
    }
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return db.ref(`users/${uid}/expenses/${id}`).update(updates).then( () => {
            dispatch(editExpense(id, updates))
        }).catch((e) => {
            console.log('Unable to edit expense', e);
        });
    }
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return db.ref(`users/${uid}/expenses`).once('value').then( (snapshot) => {
            const expenses = [];
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            })

            dispatch(setExpenses(expenses))
        }).catch((e) => {
            console.log('Unable to return expenses', e);
        });
    }
};
