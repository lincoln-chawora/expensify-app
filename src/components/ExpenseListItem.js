import React from "react";
import {Link} from "react-router-dom";
import moment from "moment";

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
    <div>
        <h3><Link to={"/edit/" + id}>{description}</Link></h3>
        <p>{amount} - {moment(createdAt).format('Do MMMM YYYY')}</p>
    </div>
)

export default ExpenseListItem;