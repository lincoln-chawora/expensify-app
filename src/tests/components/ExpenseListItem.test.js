import React from "react";
import {shallow} from "enzyme";
import expenses from "../fixtures/expenses";
import ExpenseListItem from "../../components/ExpenseListItem";

test('Should render single expense item', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]} />);
    expect(wrapper).toMatchSnapshot();
});