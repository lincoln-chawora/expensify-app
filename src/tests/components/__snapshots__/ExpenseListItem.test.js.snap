// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render single expense item 1`] = `
<div>
  <h3>
    <Link
      to="/edit/1"
    >
      Gum
    </Link>
  </h3>
  <p>
    Â£1.95
     - 
    1st January 1970
  </p>
</div>
`;
